official/rnn_ptb/util.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utility functions that perform various calculations."""
+
+import collections
+
+import tensorflow as tf
+
+EOS = '<eos>'  # End of sequence symbol
+UNK = '<unk>'  # Unknown symbol
+N = 'N'
+DS = '$'
+
+SPECIAL_SYMBOLS = [UNK, N, DS, EOS]
+
+
+def build_vocab_id_dict(data_file):
+  """Create a dictionary mapping all words in the vocabulary to integer IDs."""
+  assert tf.gfile.Exists(data_file), 'File %s does not exist.' % data_file
+  with tf.gfile.Open(data_file, 'r') as f:
+    unique_words = set(f.read().split())
+
+    # Remove special symbols from the set.
+    for symbol in SPECIAL_SYMBOLS:
+      if symbol in unique_words:
+        unique_words.remove(symbol)
+    # Create list of unique words with the special symbols at the front.
+    unique_words = SPECIAL_SYMBOLS + sorted(list(unique_words))
+
+    # Create a default dict of word->id so that any words not in the dictionary
+    # is mapped to 0 (corresponding to '<unk>').
+    vocab_dict = collections.defaultdict(
+        int, zip(unique_words, range(len(unique_words))))
+    return vocab_dict
+
+
+def build_reverse_vocab_dict(vocab_dict):
+  """Build a dictionary of id->word."""
+  reverse_dict = {i: word for word, i in vocab_dict.iteritems()}
+  reverse_dict[3] = '.\n'  # Replace '<eos>' with a period with a new line.
```

Comment:
Why append a period here? Seems like this could lead to unexpected behavior if we necessarily add punctuation. Also, hard-coding `3` here might be fragile if someone adds/changes the SPECIAL_SYMBOLS, which seems like a reasonable thing to do. Maybe at least do a quick lookup of the index of EOS in SPECIAL_SYMBOLS?

Reply:
I've changed this to replace `<eos>` with the NEWLINE symbol, and added a lookup for the index of EOS. I agree it's better with less hardcoding.